app/src/main/java/one/mixin/android/util/backup/Backup.kt

```diff
@@ -471,52 +503,71 @@ private fun internalCheckAccessBackupDirectory(context: Context, uri: Uri): Bool
 }
 
 private fun copyFileToDirectory(file: File, dir: DocumentFile) {
+    if (file.isFile && file.length() > 0) {
+        val documentFile =
+            dir.createFile("application/octet-stream", file.name)
+                ?: return
+        if (documentFile.name != file.name) {
+            documentFile.delete()
+            return
+        }
+        if (file.length() == documentFile.length()) {
+            return
+        }
+        val inputStream =
+            FileInputStream(file)
+
+        documentFile.uri.copyFromInputStream(inputStream)
+    }
+}
+
+private fun copyDirectoryToDirectory(file: File, dir: DocumentFile) {
     if (file.isDirectory) {
-        val childDir = dir.findFile(file.name).run {
-            if (this?.exists() == true) {
-                this
-            } else {
-                dir.createDirectory(file.name)
-            }
-        } ?: return
-        file.listFiles()?.forEach { f ->
-            copyFileToDirectory(f, childDir)
+        val childDir = dir.createDirectory(file.name) ?: return
+        if (childDir.name != file.name) {
+            childDir.delete()
+            return
         }
-    } else {
-        val documentFile = dir.findFile(file.name).run {
-            if (this?.exists() == true) {
-                this
+        file.listFiles()?.forEach { f ->
+            if (f.isFile) {
+                copyFileToDirectory(f, childDir)
             } else {
-                dir.createFile("application/octet-stream", file.name)
+                copyDirectoryToDirectory(f, childDir)
             }
-        } ?: return
-        val inputStream = FileInputStream(file)
-        documentFile.uri.copyFromInputStream(inputStream)
+        }
     }
 }
 
 private fun copyFileToDirectory(context: Context, sourceFile: DocumentFile, parentDir: File) {
-    try {
-        val fileName = sourceFile.name ?: return
-        if (!sourceFile.exists()) {
-            return
+    val fileName = sourceFile.name ?: return
+    if (!sourceFile.exists()) {
+        return
+    }
+    val targetFile = File(parentDir, fileName)
+    if (sourceFile.isFile) {
+        if (!targetFile.exists()) {
+            val inputStream = context.contentResolver.openInputStream(sourceFile.uri) ?: return
+            targetFile.toUri().copyFromInputStream(inputStream)
         }
-        val targetFile = File(parentDir, fileName)
-        if (sourceFile.isDirectory) {
-            if (!targetFile.exists()) {
-                targetFile.mkdirs()
-            }
-            sourceFile.listFiles().forEach {
+    }
+}
+
+private fun copyDirectoryToDirectory(context: Context, sourceFile: DocumentFile, parentDir: File) {
+    val fileName = sourceFile.name ?: return
+    if (!sourceFile.exists()) {
+        return
+    }
+    val targetFile = File(parentDir, fileName)
+    if (sourceFile.isDirectory) {
+        if (!targetFile.exists()) {
+            targetFile.mkdirs()
+        }
+        sourceFile.listFiles().forEach {
```

Comment:
perhaps converting file lists to a sequence here could reduce memory usage as there might be too many files inside directories.

Reply:
> perhaps converting file lists to a sequence here could reduce memory usage as there might be too many files inside directories.

ignore this because the sequence's `foreach` is terminal.

Reply:
Yes, there won't be too much nesting here, but we can customize a iterator to reduce the memory usage of the collection.